plugins/redfish/fu-redfish-legacy-device.c

```diff
@@ -72,6 +107,214 @@ fu_redfish_legacy_device_attach(FuDevice *dev, FuProgress *progress, GError **er
+	JsonObject *json_obj;
+	const gchar *message = "Unknown failure";
+	const gchar *state_tmp;
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new(backend);
+
+	/* create URI and poll */
+	if (!fu_redfish_request_perform(request,
+					"/redfish/v1/UpdateService",
+					FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON,
+					error))
+		return FALSE;
+
+	/* percentage is optional */
+	json_obj = fu_redfish_request_get_json_object(request);
+
+	if (json_object_has_member(json_obj, "Oem")) {
+		JsonObject *oem = json_object_get_object_member(json_obj, "Oem");
+		if (oem != NULL && json_object_has_member(oem, "Hpe")) {
+			JsonObject *hpe = json_object_get_object_member(oem, "Hpe");
+			const gchar *status = json_object_get_string_member(hpe, "State");
+			/* if we are in an idle-ish state, we can proceed */
+			if (g_strcmp0(status, "Idle") == 0 || g_strcmp0(status, "Error") == 0 ||
+			    g_strcmp0(status, "Complete") == 0) {
+				return TRUE;
+			} else {
+				g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_BUSY, "device is busy");
+				return FALSE;
+			}
+		}
+	}
+
+	g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "device unsupported");
+
+	return FALSE;
+}
+
+static gboolean
+fu_redfish_legacy_device_hpe_poll_task_once(FuRedfishDevice *self,
+					    FuRedfishDevicePollCtx *ctx,
+					    GError **error)
+{
+	FuRedfishBackend *backend = fu_redfish_device_get_backend(FU_REDFISH_DEVICE(self));
+	JsonObject *json_obj;
+	const gchar *message = "Unknown failure";
+	const gchar *state_tmp;
+	g_autoptr(FuRedfishRequest) request = fu_redfish_backend_request_new(backend);
+
+	/* create URI and poll */
+	if (!fu_redfish_request_perform(request,
+					"/redfish/v1/UpdateService",
+					FU_REDFISH_REQUEST_PERFORM_FLAG_LOAD_JSON,
+					error))
+		// If we cannot poll the BMC, let's try again, it's probably updating itself
+		return TRUE;
+
+	/* percentage is optional */
+	json_obj = fu_redfish_request_get_json_object(request);
+
+	if (json_object_has_member(json_obj, "Oem")) {
+		JsonObject *oem = json_object_get_object_member(json_obj, "Oem");
+		if (oem != NULL && json_object_has_member(oem, "Hpe")) {
+			JsonObject *hpe = json_object_get_object_member(oem, "Hpe");
+
+			const gchar *status = json_object_get_string_member(hpe, "State");
+			if (g_strcmp0(status, "Error") == 0) {
+				if (json_object_has_member(hpe, "Result")) {
+					JsonObject *result =
+					    json_object_get_object_member(hpe, "Result");
+					const gchar *message_id = NULL;
+
+					if (json_object_has_member(result, "MessageId"))
+						message_id =
+						    json_object_get_string_member(result,
+										  "MessageId");
+					if (json_object_has_member(result, "Message"))
+						message = json_object_get_string_member(result,
+											"Message");
+					/* use the message */
+					g_debug("message [%s]: %s", message_id, message);
+					fu_redfish_device_poll_set_message_id(self,
+									      ctx,
+									      message_id);
+					return FALSE;
```

Comment:
shouldn't you be setting `error` here?

Reply:
Probably not, this will be done by the `fu_redfish_device_poll_set_message_id` function (hopefully).
I will add a default value in case no message matched.